### PR TITLE
ui: modify content path

### PR DESCRIPTION
**Pull request by s-christoff**
Allows user to replace `/ui/` with their own path as long as it contains alphanumeric, -, _, or /. 
Fixes issue #1930 , however you cannot specify `v1` as that conflicts with the API endpoint. 
At this time, you cannot specify a slash to strip the ending prefix or to be used with a subdomain.

### CODE REVIEW

File: agent/config/builder.go [hunk 1]
```diff
@@ -905,6 +906,8 @@ func (b *Builder) Validate(rt RuntimeConfig) error {
 	// reDatacenter defines a regexp for a valid datacenter name
 	var reDatacenter = regexp.MustCompile("^[a-z0-9_-]+$")
 
+	// validContentPath defines a regexp for a valid content path name.
+	var validContentPath = regexp.MustCompile(`[^(?!v1)\/A-Za-z0-9\\_\\-]+`)
```
Comment (rboyer) [inline agent/config/builder.go]: Should we future proof this and prevent folks from using `v[0-9]+`?

File: agent/config/config.go [hunk 2]
```diff
@@ -264,6 +264,7 @@ type Config struct {
 	Telemetry                        Telemetry                `json:"telemetry,omitempty" hcl:"telemetry" mapstructure:"telemetry"`
 	TranslateWANAddrs                *bool                    `json:"translate_wan_addrs,omitempty" hcl:"translate_wan_addrs" mapstructure:"translate_wan_addrs"`
 	UI                               *bool                    `json:"ui,omitempty" hcl:"ui" mapstructure:"ui"`
+	UIContentPath                    *string                  `json:"uicontentpath,omitempty" hcl:"uicontentpath" mapstructure:"uicontentpath"`
```
Comment (rboyer) [inline agent/config/config.go]:
```suggestion
	UIContentPath                    *string                  `json:"ui_content_path,omitempty" hcl:"ui_content_path" mapstructure:"ui_content_path"`
```

Snake case has precedent in other config key names.

File: agent/http.go [hunk 3]
```diff
@@ -83,15 +87,82 @@ type HTTPServer struct {
...
+}
 
 type redirectFS struct {
-	fs http.FileSystem
+	fs          http.FileSystem
+	ContentPath string
 }
 
 func (fs *redirectFS) Open(name string) (http.File, error) {
 	file, err := fs.fs.Open(name)
```
Comment (rboyer) [inline agent/http.go]: Should the index.html name check come before this Open, so that you don't double-open?

File: build-support/functions/10-util.sh [hunk 4]
```diff
@@ -974,7 +974,7 @@ function ui_version {
       return 1
    fi
 
-   local ui_version="$(grep '<!-- CONSUL_VERSION: .* -->$' "$1" | sed 's/^<!-- CONSUL_VERSION: \(.*\) -->$/\1/')" || return 1
+   local ui_version="$(grep '<!-- CONSUL_VERSION: .* -->' "$1" | sed 's/<!-- CONSUL_VERSION: \(.*\) -->/\1/' | xargs)" || return 1
```
Comment (rboyer) [inline build-support/functions/10-util.sh]: Why was this change necessary?

File: agent/config/config.go (hunk 2 above)
Comment (s-christoff) [reply agent/config/config.go]: Good catch- thankyou

File: agent/http.go (hunk 3 above)
Comment (s-christoff) [reply agent/http.go]: Do you mean like:
```
	if name == "/index.html" || err != nil {
		file, err = fs.fs.Open("/index.html")
		content, _ := ioutil.ReadAll(file)
		file.Seek(0, 0)
		t, _ := template.New("fmtedindex").Parse(string(content))
		var out bytes.Buffer
		err = t.Execute(&out, fs)
		file = newTemplatedFile(&out, file)
	}else{
        file, err := fs.fs.Open(name)
}
return file,err
}
```

_Excuse the bad formatting_

File: build-support/functions/10-util.sh (hunk 4 above)
Comment (s-christoff) [reply build-support/functions/10-util.sh]: Regex wasn't matching properly anymore due to whitespacing issues. 
Ex:
```
ERROR: UI version mismatch. Expecting: 'v1.5.1-8-g33bfddb90-dev (33bfddb90)' found '  <!-- CONSUL_VERSION: v1.5.1-8-g33bfddb90-dev (33bfddb90) -->'
```

Comment (johncowen) [reply build-support/functions/10-util.sh]: Previously the things we had to print into the `index.html` file where reasonably simple.

https://github.com/hashicorp/consul/blob/master/ui-v2/lib/startup/index.js, note the single line CONSUL_VERSION.

In order to make this feature work we have to do slightly more logic and print more things out.

The templating engine that ember uses to produce `index.html` _looks_ like handlebars, but is not. It's like a 'handlebars lite', so you don't have access to all of the template helper functions and I don't think you can easily pass in variables etc.

So we moved as much as we could to use native javascript templates instead, meaning we can pass in variables and make things a lot easier to work with, which included indentation:

https://github.com/hashicorp/consul/blob/1930-mod-context-path/ui-v2/lib/startup/templates/head.html.js

Once we included indentation, we realized the grepping for the CONSUL_VERSION string stopped working because of the `^..$`, so we removed it for a looser regex.

Comment (mkeeler) [inline build-support/functions/10-util.sh]: Just curious as to why this was needed and what the actual change is besides getting rid of anchors in the regex?

File: agent/http.go (hunk 3 above)
Comment (mkeeler) [reply agent/http.go]: We should prevent double open but I think I proposed a better way overall down in one of my comments.

Create a `templatedIndexFS` type which all it does is Opens the file and if its name was index.html templates it. The redirectFS would wrap that and make the logic of each Open much cleaner.

**Comment by mkeeler** [comment]
@s-christoff Looks like you need to run `go fmt` on agent/http.go according to circle

**Comment by i0rek** [comment]
Great set of changes - looking forward to having context path finally! 🎉

File: website/source/docs/agent/options.html.md [hunk 5]
```diff
@@ -457,6 +457,8 @@ will exit with an error at startup.
   the Web UI resources for Consul. This will automatically enable the Web UI. The directory must be
   readable to the agent. Starting with Consul version 0.7.0 and later, the Web UI assets are included in the binary so this flag is no longer necessary; specifying only the `-ui` flag is enough to enable the Web UI. Specifying both the '-ui' and '-ui-dir' flags will result in an error.
 
+* <a name="_ui_content_path"></a><a href="#_ui_content_path">`-ui-content-path`</a> - This flag provides the option to change the path the Web UI loads from. By default this will go to `/ui/`. Alphanumerics, -, and _, only, `/v1/` is not able to be as it would overwrite the API endpoint.
```
Comment (i0rek) [inline website/source/docs/agent/options.html.md]:
```suggestion
* <a name="_ui_content_path"></a><a href="#_ui_content_path">`-ui-content-path`</a> - This flag provides the option to change the path the Web UI loads from. By default this will go to `/ui/`. Alphanumerics, -, and _, only, `/v1/` is not allowed as it would overwrite the API endpoint. 
```

File: website/source/docs/agent/options.html.md [hunk 6]
```diff
@@ -457,6 +457,8 @@ will exit with an error at startup.
   the Web UI resources for Consul. This will automatically enable the Web UI. The directory must be
   readable to the agent. Starting with Consul version 0.7.0 and later, the Web UI assets are included in the binary so this flag is no longer necessary; specifying only the `-ui` flag is enough to enable the Web UI. Specifying both the '-ui' and '-ui-dir' flags will result in an error.
 
+* <a name="_ui_content_path"></a><a href="#_ui_content_path">`-ui-content-path`</a> - This flag provides the option to change the path the Web UI loads from. By default this will go to `/ui/`. Alphanumerics, -, and _, only, `/v1/` is not allowed as it would overwrite the API endpoint.
```
Comment (kaitlincarter-hc) [inline website/source/docs/agent/options.html.md]:
```suggestion
* <a name="_ui_content_path"></a><a href="#_ui_content_path">`-ui-content-path`</a> - This flag provides the option to change the path the Consul UI loads from and will be displayed in the browser. By default, the path is `/ui/`, for example `http://localhost:8500/ui/`. Only alphanumerics, `-`, and `_` are allowed in a custom path. `/v1/` is not allowed as it would overwrite the API endpoint. 
```

Comment (kaitlincarter-hc) [reply website/source/docs/agent/options.html.md]: I've been trying to update instances of `Web UI` to `Consul UI`. I added some additional context to make the purpose of the flag clearer. My only hesitation is that it sounds like you are loading something from the server, "the path Consul UI loads from". If that's not the case we should revisit the language.

Comment (s-christoff) [reply website/source/docs/agent/options.html.md]: LGTM

File: agent/http.go [hunk 7]
```diff
@@ -83,9 +87,70 @@ type HTTPServer struct {
...
+}
+
+func (t *templatedFile) Sys() interface{} {
+	return nil
+}
 
 type redirectFS struct {
-	fs http.FileSystem
+	fs          http.FileSystem
+	ContentPath string
```
Comment (mkeeler) [inline agent/http.go]: You can get rid of the ContentPath here
